### PR TITLE
[Do not merge] Testing "respect memory model" PR

### PR DESCRIPTION
**Pull request by hrkrshnn**
The PR uses an SMT solver to see to check for writes outside of `[0, 64)` and allocations from `mload(freeMemPtr)`. 

The value of `freeMemPtr` was changed to `256`. This way, if there is a write to `[64, 256)`, the solver will throw (i.e., failing CI soltest cases.) The last three commits are the code. The first two commits are extracted from MemoryStoreRemover.

Changed to testing develop.

### CODE REVIEW

**Comment by hrkrshnn** [comment]
I ran this on current develop and semantic tests and found the following tests (only semantictests were checked)

1. semanticTests/inlineAssembly/inline_assembly_embedded_function_call.sol
2. semanticTests/inlineAssembly/inline_assembly_function_call.sol
3. semanticTests/inlineAssembly/inline_assembly_function_call2.sol
4. semanticTests/inlineAssembly/inline_assembly_function_call_assignment.sol
5. semanticTests/revertStrings/empty_v2.sol
6. semanticTests/inlineAssembly/inline_assembly_embedded_function_call.sol   
7. semanticTests/revertStrings/called_contract_has_code.sol
8. semanticTests/revertStrings/function_entry_checks_v1_abiv2.sol
9. semanticTests/revertStrings/array_slices.sol
10. semanticTests/revertStrings/bubble.sol
11. semanticTests/revertStrings/calldata_array_dynamic_invalid.sol
12. semanticTests/revertStrings/calldata_array_dynamic_static_short_decode.sol
13. semanticTests/revertStrings/calldata_array_dynamic_static_short_reencode.solAssert
14. semanticTests/revertStrings/calldata_array_invalid_length.sol
15. semanticTests/revertStrings/calldata_arrays_too_large.sol
16. semanticTests/revertStrings/calldata_tail_short.sol
17. semanticTests/revertStrings/ether_non_payable_function.sol
18. semanticTests/revertStrings/library_non_view_call.sol
19. semanticTests/revertStrings/short_input_array.sol
20. semanticTests/revertStrings/short_input_bytes.sol
21. semanticTests/revertStrings/transfer.sol
22. semanticTests/revertStrings/unknown_sig_no_fallback.sol
23. semanticTests/tryCatch/malformed_error.sol
24. semanticTests/viaYul/empty_return_corrupted_free_memory_pointer.sol 	
25. semanticTests/array/concat/bytes_concat_different_types.sol

**Comment by axic** [comment]
There are a lot of failures here:
```
UNCAUGHT EXCEPTION:
- function: "auto solidity::frontend::CompilerContext::appendInlineAssembly(const std::string &, const vector<std::string> &, const set<std::string> &, bool, const solidity::frontend::OptimiserSettings &, std::string)::(anonymous class)::operator()(const std::string &) const"
- file: CompilerContext.cpp
- line: 475


EXCEPTION STACK TRACE: --------------
/solidity/libsolidity/codegen/CompilerContext.cpp(475): Throw in function auto solidity::frontend::CompilerContext::appendInlineAssembly(const std::string &, const vector<std::string> &, const set<std::string> &, bool, const solidity::frontend::OptimiserSettings &, std::string)::(anonymous class)::operator()(const std::string &) const
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Error parsing/analyzing inline assembly block:
Invalid assembly generated by code generator.
------------------ Input: -----------------
{
						condition := or(
							gt(array_length, 0x100000000),
							gt(add(array_data_start, mul(array_length, 32 ), input_end)
					}
------------------ Errors: ----------------
Error: Expected ',' but got '}'

-------------------------------------------
```

Any idea why?

**Comment by hrkrshnn** [comment]
@axic That was from the original PR https://github.com/ethereum/solidity/pull/11106. I'll update this once CI goes green there.

File: libyul/optimiser/CatchMemoryViolators.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		m_solver->addAssertion(memoryLocation < p + n);
+		yulAssert(
+			m_solver->check({}).first != CheckResult::SATISFIABLE,
+			"Memory Violator Found! At " + AsmPrinter{m_dialect}(_functionCall)
+		);
+		m_solver->pop();
+	};
+
+	if (SemanticInformation::memory(*instruction) == SemanticInformation::Write)
+		switch(*instruction)
```
Comment (stackenbotten) [inline libyul/optimiser/CatchMemoryViolators.cpp:135]: Coding style error

**Comment by hrkrshnn** [comment]
Found the following test cases after running on develop:

```
semanticTests/array/concat/bytes_concat_different_types.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_832, _831)

semanticTests/viaYul/empty_return_corrupted_free_memory_pointer.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_6, _5)

semanticTests/tryCatch/try_catch_library_call.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At returndatacopy(_404, _403, _402)

semanticTests/tryCatch/create.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At returndatacopy(_341, _340, _339)

semanticTests/tryCatch/structured.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At returndatacopy(_390, _389, _388)

semanticTests/tryCatch/panic.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At returndatacopy(_461, _460, _459)

semanticTests/tryCatch/malformed_panic_3.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At returndatacopy(_439, _438, _437)

semanticTests/tryCatch/malformed_panic_2.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At returndatacopy(_408, _407, _406)

semanticTests/tryCatch/malformed_error.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_713, _712)

semanticTests/tryCatch/invalid_error_encoding.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At returndatacopy(_1450, _1449, _1448)

semanticTests/revertStrings/unknown_sig_no_fallback.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_43, _42)

semanticTests/revertStrings/transfer.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_67, _66)

semanticTests/revertStrings/short_input_bytes.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_140, _139)

semanticTests/revertStrings/short_input_array.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_172, _171)

semanticTests/revertStrings/library_non_view_call.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_104, _103)

semanticTests/revertStrings/ether_non_payable_function.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_100, _99)

semanticTests/revertStrings/calldata_tail_short.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_195, _194)

semanticTests/revertStrings/calldata_arrays_too_large.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_150, _149)

semanticTests/revertStrings/calldata_array_invalid_length.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_201, _200)

semanticTests/revertStrings/calldata_array_dynamic_static_short_reencode.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_364, _363)

semanticTests/revertStrings/calldata_array_dynamic_static_short_decode.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_159, _158)

semanticTests/revertStrings/calldata_array_dynamic_invalid.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_134, _133)

semanticTests/revertStrings/bubble.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_122, _121)

semanticTests/revertStrings/array_slices.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_183, _182)

semanticTests/revertStrings/function_entry_checks_v2.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_115, _114)

semanticTests/revertStrings/function_entry_checks_v1_abiv2.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_45, _44)

semanticTests/revertStrings/empty_v2.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_158, _157)

semanticTests/revertStrings/called_contract_has_code.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_184, _183)

semanticTests/inlineAssembly/inline_assembly_function_call_assignment.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_30, c1_14)

semantic!Tests/inlineAssembly/inline_assembly_function_call2.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_26, c1)

semanticTests/inlineAssembly/inline_assembly_function_call.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_24, c1)

semanticTests/inlineAssembly/inline_assembly_embedded_function_call.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_27, c1)
```

**Comment by hrkrshnn** [comment]
After rebasing on https://github.com/ethereum/solidity/pull/11106 and https://github.com/ethereum/solidity/pull/11339

the following errors remains:


```
semanticTests/tryCatch/malformed_error.sol: Exception during test: /solidity/libyul/optimiser/CatchMemoryViolators.cpp(127): Throw in function solidity::yul::CatchMemoryViolators::operator()(const solidity::yul::FunctionCall&)::<lambda(const auto:24&, const auto:25&)> [with auto:24 = solidity::smtutil::Expression; auto:25 = solidity::smtutil::Expression]
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
std::exception::what: Memory Violator Found! At mstore(_651, _650)
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_651, _650)
```
(there is a direct `mstore(0x44, ...)` and `mstore(0x24, ...)` inline assembly and commenting them makes the test pass.

```
semanticTests/inlineAssembly/inline_assembly_function_call_assignment.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_30, c1_14)

semantic!Tests/inlineAssembly/inline_assembly_function_call2.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_26, c1)

semanticTests/inlineAssembly/inline_assembly_function_call.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_24, c1)

semanticTests/inlineAssembly/inline_assembly_embedded_function_call.sol:
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
[solidity::util::tag_comment*] = Memory Violator Found! At mstore(_27, c1)
```

These also assigns inside inline assembly.

Also

```
semanticTests/viaYul/empty_return_corrupted_free_memory_pointer.sol: 
Exception during test: /solidity/libyul/optimiser/CatchMemoryViolators.cpp(127): Throw in function solidity::yul::CatchMemoryViolators::operator()(const solidity::yul::FunctionCall&)::<lambda(const auto:24&, const auto:25&)> [with auto:24 = solidity::smtutil::Expression; auto:25 = solidity::smtutil::Expression]
Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
std::exception::what: Memory Violator Found! At mstore(_6, _5)
```

**Comment by hrkrshnn** [comment]
Closing this, since the corresponding PRs were merged.

**Pull request closed without merging**